designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
+1. Remove `--rulesdir`
+1. Create a `@eslint/config` package to help users merge configs
+
+### The `.eslint.js` File
+
+The `.eslint.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 
+
+```js
+exports.config = {
+    files: "*.js",
+    ignore: "*.test.js",
+    globals: {},
+    settings: {},
+    processor: object,
+    parser: object,
+    parserOptions: {},
+    ruledefs: {},
+    rules: {}
+};
+```
+
+The following keys are new to the `.eslint.js` format:
+
+* `files` - **Required.** Determines the global pattern that this configuration applies to and also is used as the default pattern to search for when ESLint is executed on the command line.
+* `ignore` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this glob pattern are subtracted from the files specified in `files`.
+* `ruledefs` - Contains definitions for rules grouped by a specific name. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.
+
+The following keys are specified the same as in `.eslintrc` files:
+
+* `globals`
+* `settings`
+* `rules`
+* `parserOptions`
+
+The following keys are specified differently than in `.eslintrc` files:
+
+* `parser` - an object in `.eslint.js` files (a string in `.eslintrc`)
+* `processor` - an object in `.eslint.js` files (a string in `.eslintrc`)
+
+Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `.eslint.js`, these are all objects, requiring users to manually specify the objects to use.
+
+The following keys are invalid in `.eslint.js`:
+
+* `env` - responsibility of the user
+* `extends` - responsibility of the user
+* `overrides` - responsibility of the user
+* `plugins` - replaced by `ruledefs`
+* `root` - always considered `true`
+
+Each of these keys represent different ways of augmenting how configuration is calculated and all of that responsibility now falls on the user.
+
+#### Referencing Plugin Rules
+
+The `plugins` key in `.eslintrc` was an array of strings indicating the plugins to load, allowing you to specify processors, rules, etc., by referencing the name of the plugin. It's no longer necessary to indicate the plugins to load because that is done directly in the `.eslint.js` file. For example, consider this `.eslintrc` file:
+
+```yaml
+plugins:
+  - react
+
+rules:
+  react/jsx-uses-react: error
+```
+
+This file tells ESLint to load `eslint-plugin-react` and then configure a rule from that plugin. The `react/` is automatically preprended to the rule by ESLint for easy reference.
+
+In `.eslint.js`, the same configuration is achieved using a `ruledefs` key:
+
+```js
+const reactPlugin = require("eslint-plugin-react");
+
+exports.config = {
+    ruledefs: {
+        react: reactPlugin.rules
```

Comment:
One thing to consider here is how the users will be able to search the internet for specific error they receive if they use shareable config and config author decided to rename the plugin. I.e.:

```js
const reactPlugin = require("eslint-plugin-react");

exports.config = {
  ruledefs: {
    "not-angular": reactPlugin.rules
  },
  rules: {
    "not-angular/jsx-uses-react": "error"
  }
}
```
As far as I understand, with configuration like that, error on the console will be reported as `not-angular/jsx-uses-react` which will make it hard to find which plugin error came from. Also, for overriding rules, you have to know what the author of shareable config decided to name plugins as well, which means reading shareable config source code.

Reply:
Yes, there is no more canonical name for a rule in this proposal. I'm not entirely sure that matters, though. I anticipate a lot of users would just use the plugin name as their namespace. We could encourage plugin authors to export a `namespace` property so there is a canonical name to refer to:

```js
const reactPlugin = require("eslint-plugin-react");

exports.config = {
  ruledefs: {
    [reactPlugin.namespace]: reactPlugin.rules
  },
  rules: {
    "not-angular/jsx-uses-react": "error"
  }
}
```

Would that make sense?

Reply:
could maybe we just pass in `reactPlugin`, only, and let the schema of that object be a contract maintained between plugins and eslint (ie, not involving end users)?

Reply:
@ilyavolodin I think that if we help them with tools like `--init` then this becomes clearer. Eventually everyone will understand computed properties, so I'm not too worried about that.

@ljharb what type of contract are you envisioning? In this proposal, the concept for a formal plugin really goes away as you can pass any object that contains rules into `ruledefs`, so requiring plugins to look a certain way could restrict what otherwise is a wide-open playing field.

Reply:
You could start with "an object with a `rules` property", but then that allows for future backwards-compatible expansion, without requiring any config changes for users (only core changes, and then corresponding plugin changes).

Reply:
Please see the FAQ I added about this.

Reply:
I'm adding a new section explaining how we can retain canonical rule names without making changes to this design.